server/src/main/java/org/opensearch/action/admin/cluster/state/term/TransportClusterTermVersionAction.java

```diff
@@ -0,0 +1,110 @@
+ * under the License.
+ */
+
+/*
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.action.admin.cluster.state.term;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.action.support.ActionFilters;
+import org.opensearch.action.support.clustermanager.TransportClusterManagerNodeReadAction;
+import org.opensearch.cluster.ClusterState;
+import org.opensearch.cluster.block.ClusterBlockException;
+import org.opensearch.cluster.metadata.IndexNameExpressionResolver;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.inject.Inject;
+import org.opensearch.core.action.ActionListener;
+import org.opensearch.core.common.io.stream.StreamInput;
+import org.opensearch.threadpool.ThreadPool;
+import org.opensearch.transport.TransportService;
+
+import java.io.IOException;
+
+/**
+ * Transport action for obtaining cluster term and version from cluster-manager
+ *
+ * @opensearch.internal
+ */
+public class TransportClusterTermVersionAction extends TransportClusterManagerNodeReadAction<
+    ClusterTermVersionRequest,
+    ClusterTermVersionResponse> {
+
+    private final Logger logger = LogManager.getLogger(getClass());
+
+    @Inject
+    public TransportClusterTermVersionAction(
+        TransportService transportService,
+        ClusterService clusterService,
+        ThreadPool threadPool,
+        ActionFilters actionFilters,
+        IndexNameExpressionResolver indexNameExpressionResolver
+    ) {
+        super(
+            ClusterTermVersionAction.NAME,
+            false,
+            transportService,
+            clusterService,
+            threadPool,
+            actionFilters,
+            ClusterTermVersionRequest::new,
+            indexNameExpressionResolver
+        );
+    }
+
+    @Override
+    protected String executor() {
+        return ThreadPool.Names.SAME;
+    }
+
+    @Override
+    public ClusterTermVersionResponse read(StreamInput in) throws IOException {
+        return new ClusterTermVersionResponse(in);
+    }
+
+    @Override
+    protected ClusterBlockException checkBlock(ClusterTermVersionRequest request, ClusterState state) {
+        return null;
+    }
+
+    @Override
+    protected void clusterManagerOperation(
+        ClusterTermVersionRequest request,
+        ClusterState state,
+        ActionListener<ClusterTermVersionResponse> listener
+    ) throws Exception {
+        ActionListener.completeWith(listener, () -> buildResponse(request, state));
+    }
+
+    private ClusterTermVersionResponse buildResponse(ClusterTermVersionRequest request, ClusterState state) {
+        logger.trace("Serving cluster term version request using term {} and version {}", state.term(), state.version());
+        return new ClusterTermVersionResponse(state.getNodes().getClusterManagerNode(), state.term(), state.getVersion());
```

Comment:
Should we check here if the node who is serving the request is actually the cluster manager node ?
Is it possible that data node sends the request to cluster manager node, request reaches on transport thread and just after that master is switched for some reason. 

We would return the `state.term()` of current node and that many not be the active cluster manager node. I also wanted to understand how are we throwing all the exceptions of master switch what we're handling at data node level ?

Reply:
This is being handled in the existing implementation of TransportClusterManagerNodeAction#AsyncSingleAction#doStart (line #240). The node serving the request has a check that the clusterManager is the local node, if not it redirects to clusterManager.